trie/snap_difflayer.go

```diff
@@ -0,0 +1,231 @@
+// Copyright 2022 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package trie
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+// diffLayer represents a collection of modifications made to the in-memory tries
+// after running a block on top.
+//
+// The goal of a diff layer is to act as a journal, tracking recent modifications
+// made to the state, that have not yet graduated into a semi-immutable state.
+type diffLayer struct {
+	// Immutables
+	root   common.Hash                              // Root hash to which this snapshot diff belongs to
+	id     uint64                                   // Corresponding state id
+	nodes  map[common.Hash]map[string]*nodeWithPrev // Cached trie nodes indexed by owner and path
+	memory uint64                                   // Approximate guess as to how much memory we use
+
+	parent snapshot     // Parent snapshot modified by this one, never nil, **can be changed**
+	stale  bool         // Signals that the layer became stale (state progressed)
+	lock   sync.RWMutex // Lock used to protect parent and stale fields.
+}
+
+// newDiffLayer creates a new diff on top of an existing snapshot, whether that's a low
+// level persistent database or a hierarchical diff already.
+func newDiffLayer(parent snapshot, root common.Hash, id uint64, nodes map[common.Hash]map[string]*nodeWithPrev) *diffLayer {
+	dl := &diffLayer{
+		root:   root,
+		id:     id,
+		nodes:  nodes,
+		parent: parent,
+	}
+	var (
+		size  int64
+		count int
+	)
+	for _, subset := range nodes {
+		for path, n := range subset {
+			dl.memory += uint64(n.memorySize(len(path)))
+			size += int64(uint16(len(path)) + n.size)
+		}
+		count += len(subset)
+	}
+	triedbDirtyWriteMeter.Mark(size)
+	triedbDiffLayerSizeMeter.Mark(int64(dl.memory))
+	triedbDiffLayerNodesMeter.Mark(int64(count))
+	log.Debug("Created new diff layer", "id", id, "nodes", count, "size", common.StorageSize(dl.memory))
+	return dl
+}
+
+// Root returns the root hash of corresponding state.
+func (dl *diffLayer) Root() common.Hash {
+	return dl.root
+}
+
+// ID returns the state id represented by layer.
+func (dl *diffLayer) ID() uint64 {
```

Comment:
ID is corresponding to a state transition, not relevant with block number. I will introduce the differences from two aspects:

State(0) is corresponding to the initial state, which can be genesis state, or the snap sync'd state, while Block(0) is only corresponding to the genesis block.

Block may not trigger any state transition in case (1) the block has no transaction (2) there is no block reward mechanism in execution layer, e.g. clique, or post-merge network. If no state transition happens, no new layer will be created.

All in all, id is corresponding to state transition.

Reply:
And what's more, actually, for the canonical states which form a chain, these states will never be equal because of nonce bumping. We can confidently say these all canonical states are unique, so represented by a unique integer.